vm/src/obj/objint.rs

```diff
@@ -810,79 +786,154 @@ pub fn to_int(vm: &VirtualMachine, obj: &PyObjectRef, base: &BigInt) -> PyResult
     }
 }
 
-fn str_to_int(literal: &str, mut base: u32) -> Option<BigInt> {
-    let mut buf = validate_literal(literal)?.to_owned();
-    let is_signed = buf.starts_with('+') || buf.starts_with('-');
-    let radix_range = if is_signed { 1..3 } else { 0..2 };
-    let radix_candidate = buf.get(radix_range.clone());
-
-    // try to find base
-    if let Some(radix_candidate) = radix_candidate {
-        if let Some(matched_radix) = detect_base(&radix_candidate) {
-            if base == 0 || base == matched_radix {
-                /* If base is 0 or equal radix number, it means radix is validate
-                 * So change base to radix number and remove radix from literal
-                 */
-                base = matched_radix;
-                buf.drain(radix_range);
-
-                /* first underscore with radix is validate
-                 * e.g : int(`0x_1`, base=0) = int('1', base=16)
-                 */
-                if buf.starts_with('_') {
-                    buf.remove(0);
+fn to_int_radix(vm: &VirtualMachine, obj: &PyObjectRef, base: u32) -> PyResult<BigInt> {
+    debug_assert!(base == 0 || (2..=36).contains(&base));
+
+    let opt = match_class!(match obj.clone() {
+        string @ PyString => {
+            let s = string.as_str();
+            bytes_to_int(s.as_bytes(), base)
+        }
+        bytes @ PyBytes => {
+            let bytes = bytes.get_value();
+            bytes_to_int(bytes, base)
+        }
+        bytearray @ PyByteArray => {
+            let inner = bytearray.borrow_value();
+            bytes_to_int(&inner.elements, base)
+        }
+        _ => {
+            return Err(
+                vm.new_type_error("int() can't convert non-string with explicit base".to_owned())
+            );
+        }
+    });
+    match opt {
+        Some(int) => Ok(int),
+        None => Err(vm.new_value_error(format!(
+            "invalid literal for int() with base {}: {}",
+            base,
+            vm.to_repr(obj)?,
+        ))),
+    }
+}
+
+fn bytes_to_int(lit: &[u8], mut base: u32) -> Option<BigInt> {
+    // split sign
+    let mut lit = lit.trim();
+    let sign = match lit.first()? {
+        b'+' => Some(Sign::Plus),
+        b'-' => Some(Sign::Minus),
+        _ => None,
+    };
+    if sign.is_some() {
+        lit = unsafe {
+            // safe by sign check
+            lit.get_unchecked(1..)
+        };
```

Comment:
This was fine as well as it was

Reply:
yes, it seems `lit.first()?` is covered well unlike `Some()` branch

Reply:
Actually, I think using `get_unchecked` is better when we are not sure it is properly optimized or not.
Still going to this way because I think this project is not on the performance optimization phase.